lib/tree.rb

```diff
@@ -1,64 +1,177 @@
+    temp = TreeNode.new(key, value) 
+    
+    if @root == nil 
+      @root = temp
+      return
+      
+    else
+      add_helper(@root, temp)
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(current_node, temp)
+    if current_node == nil
+      return temp
+    end
+    
+    if temp.key < current_node.key 
+      current_node.left = add_helper(current_node.left, temp)
+    else
+      current_node.right = add_helper(current_node.right, temp)
+    end
+    return current_node
+  end
+  
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
   def find(key)
-    raise NotImplementedError
+    current = @root
+    if current == nil
+      return nil
+    elsif current.key == key
+      return current.value
+    else
+      find_helper(current, key)
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def find_helper (current, key) 
+    if current.key == key
+      return current.value
+    elsif current.key > key && current.left != nil 
+      find_helper(current.left, key)
+    elsif current.key > key
+      return nil
+    elsif current.key < key && current.right != nil
+      find_helper(current.right, key)
+    elsif current.key < key
+      return nil
+    end
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    if @root == nil
+      return []
+    end
+    inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(current_node, list)
+    
+    if current_node == nil
+      return list
+    end
+    
+    inorder_helper(current_node.left, list)
+    list.push({key: current_node.key, value: current_node.value})
+    inorder_helper(current_node.right, list)
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    if @root == nil
+      return []
+    else
+      return preorder_helper(@root, [])
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(current_node, list)
+    if current_node == nil
+      return list
+    end
+    
+    list.push({key: current_node.key, value: current_node.value})
+    preorder_helper(current_node.left, list)
+    preorder_helper(current_node.right, list)
+    
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    if @root == nil
+      return []
+    else
+      return postorder_helper(@root, [])
+    end
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(current_node, list)
+    if current_node == nil
+      return list
+    end
+    
+    postorder_helper(current_node.left, list)
+    postorder_helper(current_node.right, list)
+    list.push({key: current_node.key, value: current_node.value})
+    
+    return list
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
You're visiting each node so the time complexity is O(n), but because the call stack doesn't go deeper than the height of the tree the space complexity is O(log n) for a balanced tree and O(n) for an unbalanced one.